[16.0][ADD] cetmix_tower_server-implement_vault_for_secrets

halbtonjazz opened the pull request:
Implement a new approach for storing confidential data in a centralized
way to avoid code duplication and improve maintainability.

Currently we have several occasions in several models where we store 
confidential information. This means that we have the same code 
duplicated several times which makes code maintenance less convenient 
and is also error prone.

Changes include:
- Add new abstract model cx.tower.vault.mixin to handle secret values
- Add new model cx.tower.vault to store the secret values  
- Apply the mixin to cx.tower.key, cx.tower.key.value, and
  cx.tower.server models
- Move secret handling logic from individual models to the mixin
- Set proper security rules for the vault model

This is a draft implementation that provides centralized secret storage
mechanism which can be easily extended to use external key management
systems by overriding the mixin functions.

Task: 4824

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- New Features
  - Secure, vault-backed storage for sensitive fields (passwords, keys) across servers and keys.
  - Secrets are redacted with a placeholder in views and command outputs, with controlled retrieval when needed.
  - Creation-time validation ensures SSH password is provided when using password-based authentication.

- UI
  - Broader menu visibility by removing group restrictions from key sections.
  - Tools menu repositioned under the root menu; Settings order adjusted.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

halbtonjazz committed:
[ADD] cetmix_tower_server: implement vault for secrets

Implement a new approach for storing confidential data in a centralized
way to avoid code duplication and improve maintainability.

The solution introduces a centralized secret storage mechanism through
an abstract mixin and dedicated vault model. This approach eliminates
code duplication across cx.tower.key, cx.tower.key.value, and
cx.tower.server models by moving shared secret handling logic to a
reusable component.

This implementation provides a foundation that can be easily extended
to use external key management systems by overriding the mixin
functions, improving both security and scalability.

Task: 4824

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,191 @@
+        """
+        domain = [
+            ("res_model", "=", self._name),
+            ("res_id", "=", self.id),
```
What if `self` contains multiple records?

halbtonjazz replied:
I have resolved this issue in a new commit. Please check it

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,191 @@
+
+            domain = [
+                ("res_model", "=", self._name),
+                ("res_id", "=", self.id),
```
Same here

halbtonjazz replied:
I have resolved this issue in a new commit. Please check it

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,191 @@
+                ("res_id", "=", self.id),
+                ("field_name", "=", field),
+            ]
+            vaults_data = self.env["cx.tower.vault"].sudo().search(domain)
```
Running SQL query inside a loop is really a bad idea in terms of performance. You should fetch all the data outside the look and use helper functions such as `filtered` afterwards.

halbtonjazz replied:
Please review the update

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,191 @@
+            if vaults_data:
+                vaults_data.write({"data": value})
+            else:
+                self.env["cx.tower.vault"].sudo().create(
```
Same for `create`. This function supports val_list, so you should prepare a list inside the loop and than call the `create` once.

halbtonjazz replied:
Please review the update

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,191 @@
+                    secret_values[secret_field] = vals.pop(secret_field)
+
+            # Create record without secret fields
+            new_record = super().create(vals)
```
Why not to sanitise the vals and create all the records afterwards?

halbtonjazz replied:
Yes, you're right, thanks for the suggestion

halbtonjazz replied:
The problem is that we cannot compare the created records with the original vals because there is no identifier that can link them together, and the sorting in the model where cx.tower.vault.mixin will be inherited may be different.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,230 @@
+    SECRET_VALUE_PLACEHOLDER = "*** Insert new value to replace the existing one ***"
+    SECRET_VALUE_SPOILER = "*****"
+    SECRET_FIELDS = []
+    _placeholder_field = "SECRET_VALUE_PLACEHOLDER"
```
Why do we need to add an extra variable instead of simply using the `SECRET_VALUE_PLACEHOLDER`?

halbtonjazz replied:
Because depending on the model, we need to use `SECRET_VALUE_SPOILER` or `SECRET_VALUE_PLACEHOLDER `(in the `_read` method, as in the current code)

ivs-cetmix replied:
These are two different things used in two different scenarios. Placeholder is always a placeholder, while spoiler is always a spoiler.
Moreover we can simply use only placeholder instead of having two different entities that are doing virtually the same thing.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,230 @@
+
+        # Group vault data by res_id
+        vault_data_by_record = {}
+        for vault_record in vault_records:
```
Why not simply to use [search_read](https://github.com/odoo/odoo/blob/c637426f676c8fd142196fc8610995ebe9ef82b1/odoo/models.py#L5038) instead?

halbtonjazz replied:
fixed

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,230 @@
+        existing_vault_records = self.env["cx.tower.vault"].sudo().search(domain)
+
+        # Prepare data for batch operations
+        records_to_update = []
```
You should opt for using reocrdset when working with recordsets, not a list.

halbtonjazz replied:
fixed

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,230 @@
+
+                if existing_record:
+                    # Prepare for update
+                    records_to_update.append((existing_record, str(value)))
```
Check the comment above

halbtonjazz replied:
fixed

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,230 @@
+                    )
+
+        # Batch operations
+        for vault_record, new_value in records_to_update:
```
Can it happen so that different records will have different values? If no, then we need to update all records at once.

halbtonjazz replied:
fixed

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,230 @@
+        records = super().create(sanitized_vals_list)
+
+        # Assign secret values to corresponding records
+        for record, secret_values in zip(records, secret_values_list, strict=True):
```
strict doesnt' work on Python 3.10 which is used in Odoo official docker image for 16.0 
Check [this](https://github.com/cetmix/cetmix-tower/blob/c7b1a8fb76090f2d04bcdb1e7809cb23cf0b1683/cetmix_tower_server/tests/test_server.py#L273)

halbtonjazz replied:
I had to use `strict=True` because pre-commit was requiring it. However, I've now rewritten the  `create` method to use `@api.model_create_single` instead of `@api.model_create_multi` since single record creation is required because batch creation doesn't allow matching values from  `vals_list` with the created records when storing secret fields in the vault. We need the exact record ID to create corresponding vault entries for each secret field

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,230 @@
+
+        return vaults_data
+
+    def _set_secret_value(self, vals):
```
Use plural to reflect the function nature -> `values`

halbtonjazz replied:
fixed

GabbasovDinar reviewed (changes requested):
@halbtonjazz please, fix tests

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,269 @@
+        # Store secret values in vault
+        if secret_values:
+            res.set_secret_values(secret_values)
+            table_name = res._table
```
Res is the same model as self, so `self._table`

ivs-cetmix replied:
You should update all the records at once, it will be just a single query instead of multiple ones.

halbtonjazz replied:
fixed

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,269 @@
+
+            secret_field_list = list(secret_values.keys())
+            for secret_field in secret_field_list:
+                query = f"UPDATE {table_name} SET {secret_field} = NULL WHERE id = %s"
```
Please add a complete comment that describes why you are updating the database table directly here.

halbtonjazz replied:
added

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,276 @@
+        self.ensure_one()
+
+        # Validate field is configured as secret field
+        if field_name not in self.SECRET_FIELDS:
```
Why not to simply call the `get_secrect_values` function and returns a singe result for the selected record?

halbtonjazz replied:
done

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,276 @@
+    # matching values from vals_list with the created records when storing
+    # secret fields in the vault. We need the exact record ID to create
+    # corresponding vault entries for each secret field.
+    @api.model_create_single
```
We must avoid single record creation, so must be redesigned to use the `create_multi`.  Probably you can store some secret ID instead of the actual value and than identify the final record using that ID before flushing the fields.

halbtonjazz replied:
done

GabbasovDinar commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,301 @@
+        records_secret_data = []  # [(temp_secret_id, secret_values), ...]
+
+        for vals in vals_list:
+            secret_values = {}
+            for secret_field in self.SECRET_FIELDS:
+                if secret_field in vals and vals.get(secret_field):
+                    secret_values[secret_field] = vals.pop(secret_field)
```
secret_vals = {f: vals.pop(f) for f in self.SECRET_FIELDS if f in vals and vals[f]}

halbtonjazz replied:
Thanks for the suggestion, cool

GabbasovDinar commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,301 @@
+        """Create records and set secret values if present"""
+
+        # Step 1: Extract secret fields and generate temporary IDs
+        records_secret_data = []  # [(temp_secret_id, secret_values), ...]
```
records_secret_data = []
vault_vals_list = []

halbtonjazz replied:
added, thank for the suggestion

GabbasovDinar commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,301 @@
+                            "field_name": field_name,
+                            "data": field_value,
+                        }
+                    )
```
```python
vault_vals_list.append({
                            "res_model": self._name,
                            "secret_id": secret_id,  # Temporary ID for linking
                            "field_name": field_name,
                            "data": field_value,
})
```

halbtonjazz replied:
fixed

GabbasovDinar commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,301 @@
+                    )
+            else:
+                records_secret_data.append((None, {}))
+
```
```python
if vault_vals_list:
    self.env["cx.tower.vault"].sudo().create(vault_vals_list)
```

use create multi instead of create one-by-one

halbtonjazz replied:
thanks for optimizing

GabbasovDinar commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,301 @@
+
+                # Clear secret fields from main table
+                secret_field_list = list(secret_values.keys())
+                record.invalidate_recordset(secret_field_list)
```
```python
    for record, (secret_id, field_names) in zip(records, records_secret_data):
        if secret_id and field_names:
            vaults = Vault.search([
                ("res_model", "=", self._name),
                ("secret_id", "=", secret_id),
                ("field_name", "in", field_names),
            ])
            vaults.write({"res_id": record.id})
            record.invalidate_recordset(field_names)
```

halbtonjazz replied:
Ok, but I intentionally used enumerate because `zip` gives a pre-commit error requiring `strict=True`. And Ivan mentioned in this PR that strict doesn't work on Python 3.10 which is used in Odoo official docker image for 16.0. But no problem, I've fixed it to use zip and added `strict=True`

GabbasovDinar commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,301 @@
+                    except Exception:  # pylint: disable=except-pass
+                        # skip SpecialValue
+                        # (e.g. for missing record or access right)
+                        pass
```
```python
show_all = not fields

secret_fields = self.SECRET_FIELDS if show_all else [f for f in self.SECRET_FIELDS if f in fields]

for record in self:
    for secret_field in secret_fields:
        try:
            ....

```

halbtonjazz replied:
fixed, thank for the suggestion

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,304 @@
+        )
+        return vault_records
+
+    def set_secret_values(self, vals):
```
Same here. Such function should be private.

halbtonjazz replied:
fixed

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,304 @@
+        # Return None if no secret value found anywhere
+        return None
+
+    def get_secret_values(self, fields_list=None):
```
Same here

halbtonjazz replied:
fixed

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,304 @@
+                    # (e.g. for missing record or access right)
+                    pass
+
+    def get_secret_value(self, field_name):
```
I think it's better to make this function private. 
@GabbasovDinar your thoughts?

halbtonjazz replied:
The question was not addressed to me, but this method is intentionally public as it's used extensively outside the class - in `cx_tower_key.py`, `cx_tower_server.py`, `tests`, etс. But if you want to add _, I'll add it, no problem

ivs-cetmix replied:
It's not used outside the class to much because `cx_tower_key.py` and `cx_tower_server.py` inherit this mixin, so all calls are inside the same class. 
Making it private makes is unavailable for the rpc calls, which can be good from the security point of view.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,304 @@
+
+            if secret_vals:
+                # Generate temporary unique ID for this record
+                secret_id = uuid.uuid4()
```
This is costly method. Besides it's not temporary because it's saved in the db.

halbtonjazz replied:
I looked at how this is implemented in odoo core and did it in a similar way. The previous implementation without resource costs was writing to the main model and then deleting

ivs-cetmix replied:
The main thing is that we don't need this data after it's stored. Which means it must be deleted at the end of the transaction.  So you can populate simple integers because those values will be isolated inside the transaction anyway.

halbtonjazz replied:
I've fixed this as we discussed and replaced UUID-based linking with temporary identifiers

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_server.py:
```diff
@@ -124,7 +127,9 @@ class CxTowerServer(models.Model):
-        string="SSH Password", groups="cetmix_tower_server.group_manager"
+        string="SSH Password",
+        groups="cetmix_tower_server.group_manager",
+        readonly=False,
```
It's `False` by default, no need to add it explicitly once again.

halbtonjazz replied:
fixed

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault.py:
```diff
@@ -0,0 +1,43 @@
+        required=True,
+        help="Name of the field that contains the secret value",
+    )
+    secret_id = fields.Char(
```
Why do we need to use another field? We already have res_model and res_id for.

halbtonjazz replied:
The `secret_id` field is needed to implement your previous suggestion about avoiding single record creation (https://github.com/cetmix/cetmix-tower/pull/347#discussion_r2219137984). During batch creation, we create vault records with `secret_id` before main records exist (no `res_id` yet), then link them afterward using this identifier

ivs-cetmix replied:
Yes, but we don't need to store that data. Not to mention, that you are creating another link to the same record. Check my comment to the mixin file below.

halbtonjazz replied:
I've removed the `secret_id` field concept and implemented the temp_identifier approach for vault mixin secret handling

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,324 @@
+        ):
+            return None
+
+        query = f"SELECT {field_name} from {self._table} WHERE id=%s"
```
Why are we fetching something from some table(s)?
Is this possible that single value is stored in some specific table while multiple values are stored in another place?

halbtonjazz replied:
We cannot store a `secret value` in the vault until a record is created. At the same time, compute and constraints work when a record is created. An example is the constraint `_constraint_ssh_settings`. It will check secret fields such as `ssh_password,` but since there will be no record in the vault at the time of creation, it will fail with an error because it cannot find the field values. For such cases, we write a temporary value for the secret field when creating the model itself, and this case is handled in the `get_secret_value` method. After successfully creating the record, directly in the create method, we clear the secret field from the model (record) itself and write the actual value to the vault. A later call to `get_secret_value` will return the actual value from the vault.

ivs-cetmix replied:
Constraint methods are called at the end of transaction. To that moment all the values are already stored.
Besides you store a placeholder, which means that the table is not empty. And the `_read` method returns the placeholder, so even at the ORM level there is a value.
Just below in L85 you are calling the method that does absolutely the same thing: queries the database. 
So why do you need to run another query which will return some irrelevant value?

halbtonjazz replied:
Constraints are called in the default _create method in models.py, at the end of which the constraints for store fields are checked 
https://github.com/odoo/odoo/blob/c1d3153ed564c0f47a201e5751740857a15d7db2/odoo/models.py#L4270 ,  https://github.com/odoo/odoo/blob/c1d3153ed564c0f47a201e5751740857a15d7db2/odoo/models.py#L1370
Our logic creates a record in the vault only after the model is created (after _create and before checking the constraints themselves), so we need to get this value from somewhere. For this purpose, in the create method, we pass a temporary value for the secret fields to vals to replace the actual secret value. And in order for the constraints to check the completeness of these fields to work, we added a query to the get_secret_value method to check this temporary value in the model itself.
After the entire create method is completed, our real value will be written to the vault, and therefore we will continue to take the value from it

ivs-cetmix replied:
Ok, but we put temporary values instead of the secret ones, so there will be no difference from the constraint point of view.
And if we are talking about `_constraint_ssh_settings`, this constraint is defined at the model level, not at the mixin one. So it should be addressed at the model level either.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,324 @@
+
+    def _extract_secret_fields(self, vals_list):
+        """Extract secret fields and replace with temporary identifiers.
+
```
Please provide a detailed docstring

halbtonjazz replied:
added

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,324 @@
+            for secret_field in self.SECRET_FIELDS:
+                if secret_field in vals and vals[secret_field]:
+                    temp_id_counter += 1
+                    temp_identifier = f"temp_secret_{temp_id_counter}"
```
`str(temp_id_counter)` will do the same job faster

halbtonjazz replied:
I disagree with this statement. f-strings are faster in Python 3.6+ because f-strings are optimized at the bytecode level and work faster than concatenation. In addition, this will reduce code readability and is a step away from modern python standards

ivs-cetmix replied:
This is not a concatenation. This is a type conversion. It will be much faster than an f-string.
Regarding readability, this makes no sense here because this is not a variable.
It is a temporary value used inside the function.

halbtonjazz replied:
Ok, if we're not talking about concatenation, no problem.
Fixed

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,324 @@
+        for record_dict in records_dict:
+            self._process_single_record_secrets(record_dict, secret_vals)
+
+    def _fetch_records_data(self, records):
```
This can be put inside _process_secret_values_after_creation as it doesn't contain any specific business logic.

halbtonjazz replied:
fixed

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,324 @@
+            record._set_secret_values(vault_vals)
+
+        # Invalidate cache
+        record.invalidate_recordset(fields_to_clear)
```
Imagine you have created 1000 records.

halbtonjazz replied:
fixed

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,324 @@
+        return self.env.cr.dictfetchall()
+
+    def _process_single_record_secrets(self, record_dict, secret_vals):
+        """Process secrets for a single record."""
```
Please provide complete docstring. Here and everywhere. And always.

halbtonjazz replied:
added

ivs-cetmix reviewed (changes requested):
Please ensure that all(!) functions have complete docstrings before submitting code for review.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -211,7 +211,23 @@ def _set_secret_values(self, vals):
     @api.model_create_multi
     def create(self, vals_list):
-        """Create records and set secret values if present"""
+        """Create records and handle secret values securely.
```
This is a core odoo method. It's already documented in the Odoo source code.

halbtonjazz replied:
fixed. I added this in response to this message because you asked to check that all(!) functions have complete in this comment https://github.com/cetmix/cetmix-tower/pull/347#pullrequestreview-3092146319
Before implementing the mixin for this method, the documentation was written here https://github.com/cetmix/cetmix-tower/blob/00d228f689eba607f5cfc4c9a4ea372e5c67a5a8/cetmix_tower_server/models/cx_tower_key.py#L76 and here https://github.com/cetmix/cetmix-tower/blob/14ead49b523c6b11beb93271b771891a267c3acd/cetmix_tower_server/models/cx_tower_key_value.py#L111 
I did it by analogy

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,418 @@
+                    pass
+
+    def _get_secret_value(self, field_name):
+        """Retrieve the actual secret value for a specific field.
```
Retrieves the actual secret value for a specific field **for a single record**.

halbtonjazz replied:
added

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_key.py:
```diff
@@ -443,15 +395,58 @@ def _set_secret_value(self, value=None):
+            This method first attempts to retrieve SSH key values from the main
+            table, then falls back to the parent vault implementation for secrets.
+        """
+        value = self._get_ssh_key_from_main_table(field_name)
```
Why are you fetching this value from the main table? 
It must be stored in the vault as any other secret value.

halbtonjazz replied:
This is redundant code from one of the previous implementations, which is now unnecessary

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,418 @@
+                  - 'field_name': Name of the secret field
+                  - 'data': The actual secret value
+
+                  Example: [{'res_model': 'cx.tower.server', 'res_id': 2,
```
This model is a mixin which is inherited by a model. This function cannot return secret values of any other model. So adding `res_model` doesn't have any practical reason. Same for `res_id`. 
You must return data in a convenient format that will allow to use them later.
Eg in L86 you have to iterate records in a loop. If you have a dict you could have done the same thing much faster.

halbtonjazz replied:
I agree about res_model. res_id should be kept because this method is called on more than one record. As for the format, it is implemented according to the Function details from the task description

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,417 @@
+        self.env.cr.execute(query, (tuple(records.ids),))
+        records_dict = self.env.cr.dictfetchall()
+
+        fields_to_invalidate = set()
```
You can simply invalidate all secret fields. This will not affect performance and will make the code simpler.

halbtonjazz replied:
fixed

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,417 @@
+
+        return fields_to_clear
+
+    def _clear_temp_values(self, fields_to_clear):
```
Same here. What if we have 1000 records? It's much easier and faster to clear all secret fields for the entire recordset without checking them and clearing selectively.

halbtonjazz replied:
fixed

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,417 @@
+        if secret_values:
+            self._set_secret_values(secret_values)
+            # Invalidate cache for all secret fields
+            for field_name in secret_values.keys():
```
20 fields = 20 invalidations of the same recordset

halbtonjazz replied:
fixed

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,417 @@
+                ["res_id", "field_name", "data"],
+            )
+        )
+        return vault_records
```
The original task stated you should return the result in the `read` function format. This is not a `read` function format because it contains `res_id` instead of `id`.
Anyway to make this function more usable we should implement a more convenient format, eg a dict. Same way as it is done [here](https://github.com/cetmix/cetmix-tower/blob/00d228f689eba607f5cfc4c9a4ea372e5c67a5a8/cetmix_tower_server/models/cx_tower_variable_mixin.py#L36).

halbtonjazz replied:
fixed

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,385 @@
+        for value, records in vals_to_update_records.items():
+            records.sudo().write({"data": value})
+
+        self.env["cx.tower.vault"].sudo().create(records_to_create)
```
```python
if records_to_create:
.....
```

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,385 @@
+            records.sudo().write({"data": value})
+
+        self.env["cx.tower.vault"].sudo().create(records_to_create)
+        records_to_unlink.sudo().unlink()
```
```python
if records_to_unlink:
....
```

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,385 @@
+                vault_vals[field_name] = secret_value
+
+        # Update database and vault if needed
+        record = self.browse(record_id)
```
We need the record only if we have the vault_vals

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_vault_mixin.py:
```diff
@@ -0,0 +1,385 @@
+    def unlink(self):
+        """Override unlink to delete vault records.
+
+        Automatically removes all associated vault records before deleting
```
What if there is an error in the unlink method and transaction is rolled back?

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_key.py:
```diff
@@ -443,15 +395,16 @@ def _set_secret_value(self, value=None):
-            self.invalidate_recordset(["secret_value"])
+            secret_fields_list = list(vals.keys())
+            set_clause = ", ".join(f"{field} = %s" for field in secret_fields_list)
+            query = f"UPDATE {self._table} SET {set_clause} WHERE id = %s"
```
Why are we using an SQL query instead of generic functions? Actually why do we need to do anything for the "k" type key? Wouldn't it be handled by the mixin?

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_server.py:
```diff
@@ -1950,3 +1905,21 @@ def _get_notification_action(
+            validation_errors = []
+            for vals in vals_list:
+                if vals.get("ssh_auth_mode") == "p" and not vals.get("ssh_password"):
+                    server_name = vals.get("name", "Unknown Server")
```
"name" is a mandatory field

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_key.py:
```diff
@@ -86,25 +89,10 @@ def create(self, vals_list):
             new_key = super().create(vals)
             if secret_value:
-                new_key._set_secret_value(secret_value)
+                new_key._set_secret_values({"secret_value": secret_value})
```
This is not needed any longer because mixin will do the job for ssh keys.
So we need to split the keys by type - ssh or secret.
For ssh keys we can simply call super.
For secrets we need to create them one by one as before while adjusting this process to the new mixin flow.

halbtonjazz replied:
I analyzed this fragment and realized I made an error. I didn't need to override the create method in this model. Because the mixin's create method already handles secret field values and calls the overridden _set_secret_values from cx.tower.key, which stores the secret value depending on the type - secrets or SSH keys

ivs-cetmix reviewed (changes requested):
Code LGTM, however you must follow the OCA coding guideline when in comes to adding functions if Python files.

ivs-cetmix committed:
[IMP] cetmix_tower_server: Menuitems

- Remove groups from menuitems.xml
- Adjust sequence to menuitems.xml

Task 4917

ivs-cetmix reviewed (approved):
Intermediate code review LGTM

CetmixGitDrone commented:
This PR has the `approved` label and has been created more than 5 days ago. It should therefore be ready to merge by a maintainer (or a PSC member if the concerned addon has no declared maintainer). 🤖
